text/0000-symbol-name-mangling-v2.md

```diff
@@ -330,48 +543,56 @@ std::iter::Chain<
-$8: ----------------------------------------------------------------------------------
-```
-
-The indices on the left are the dictionary keys. The prefixes `std`, `std::iter`, and `std::iter::Chain` all get added to the dictionary because we have not seen them before. After that we encounter `std` again. We've already seen it, so we don't add anything to the dictionary. The same goes for when we encounter `std::iter` the second time. Next we encounter `std::iter::Zip`, which we have not seen before, so it's added to the dictionary. Next we encounter `std` again (already seen, no insertion), then `std::vec` and `std::vec::IntoIter` which both generate a new entry. Next we see `std::vec::IntoIter<u32>`, the first full _type_. It generates an entry too. The second type parameter is the same as the first. No part of it introduces a new entry. After the next `>` we have completely processed `std::iter::Zip<std::vec::IntoIter<u32>, std::vec::IntoIter<u32>>`, which adds another type entry. Finally, the full `std::iter::Chain<std::iter::Zip<std::vec::IntoIter<u32>, std::vec::IntoIter<u32>>>` adds another entry.
-
-Using the dictionary above, we can compress to:
+Here's an example in order to illustrate the concept. The name
 
 ```
-std::iter::Chain<$1::Zip<$0::vec::IntoIter<u32>, $6>>
+std::iter::Chain<std::iter::Zip<std::vec::IntoIter<u32>, std::vec::IntoIter<u32>>>
 ```
 
-A couple of things to note:
-
- - The first occurrence of a dictionary entry is never substituted. We don't store the dictionary anywhere and need to be able to reconstruct it from the compressed version.
- - Longer substitutions are preferred to shorter ones. `std::iter::Chain<$1::Zip<$0::vec::IntoIter<u32>, $4::IntoIter<u32>>>` would also decompress to the original version but the compiler is supposed to always pick the longest substitution available.
-
-The mangled version of a substitution marker is `S <key - 1> _` (and `S_` for key `0`) like in the Itanium mangling. So the above definition would be mangled to:
+is mangled to the following uncompressed string. The lines below show parts
+of the mangled string that already occurred before and can thus be replaced
+by a back reference. The number of at the beginning of each span given
+the 0-based byte position of where it occurred the first time.
 
 ```
-_RN12std_a1b2c3d44iter5ChainINS0_3ZipINS_3vec8IntoIterIjEES5_EEE
+0         10        20        30        40        50        60        70        80        90
+_RINtNtC3std4iter5ChainINtNtC3std4iter3ZipINtNtC3std3vec8IntoItermEINtNtC3std3vec8IntoItermEEE
+                            7----              7----                    7----
+                          5-----------                                45---------
+                                                                    43--------------------
+                                                                   42-----------------------
```

Comment:
IIRC, my actually compression scheme uses `0` for the first byte *after* `_R` - in fact, this entire document could talk about the mangling as "the data after `_R`", and only specify in one place that it's encapsulated in `"_R" <data> ""` (`_R` prefix and no suffix).